### ISSUE TITLE
Support for flagging posts

### ISSUE THREAD

**Issue by UlfS** [matterhorn-chat/matterhorn#246]
Flagging posts for future reference is a convenient feature that I miss from using the official mattermost client.
It would be nice to have some support for that in matterhorn.

It would at least be nice if matterhorn didn't crash when something is flagged by another client:
```hs
matterhorn: JSONDecodeException {jsonDecodeExceptionMsg = "Error in $.event: Unknown websocket message: \"preferences_changed\"", jsonDecodeExceptionJSON = "{\"event\":\"preferences_changed\",\"data\":{\"preferences\":\"[{\\\"user_id\\\":\\\"...\\\",\\\"category\\\":\\\"flagged_post\\\",\\\"name\\\":\\\"...\\\",\\\"value\\\":\\\"true\\\"}]\"},\"broadcast\":{\"omit_users\":null,\"user_id\":\"...\",\"channel_id\":\"\",\"team_id\":\"\"},\"seq\":1}"}
```

**Comment by jtdaugherty** [comment]
By "flagging posts" do you mean "pinning", as in the "Pin to channel" option for posts in the web client? If so, we already have tickets for that (#213, #215). If not, let me know!

As for the crash you described, that is due to a Mattermost 3.9 feature. Right now Matterhorn does not support 3.9, so such crashes are to be expected if you run Matterhorn against an incompatible version of the server. The latest server version we support is 3.8.X. You can tell which server version the client supports by looking at the version number: matterhorn version `30802.1.0` supports server version `3.8.2`.

**Comment by UlfS** [comment]
Flagging posts is a different feature and was apparently introduced in [mattermost 3.3](https://about.mattermost.com/mattermost-3-3-golang-bot-chinese-korean-dutch-message-flags-here-and-more/). See "Message Flags".
This is a user-specific flag for any message and you can open a list of all the posts you flagged.

Out of scope of this issue:
I can live with the crash, it does not happen that often, but maybe it could be more relaxed and just log a warning though.
Does mattermost introduce a lot of breaking changes in minor versions?

**Comment by jtdaugherty** [comment]
Thanks for that link - I had never even noticed the flag icon. :)

In my opinion the crash is actually preferable to a warning, since in the presence of a protocol parse failure, we cannot make any guarantees about proper client functioning so continuing to run would be misleading. As to your question, yes: just about every Mattermost release makes breaking changes. We are actually working with the Mattermost folks to improve the situation that you encountered, by implementing https://github.com/matterhorn-chat/matterhorn/issues/203. Once that's implemented, Matterhorn will do a better job of telling users up front when it is incompatible with the server version.

**Comment by jtdaugherty** [comment]
This feature is now under development in a branch at https://github.com/matterhorn-chat/matterhorn/tree/feature/flagging.

**Comment by aisamanra** [comment]
The current state of the branch includes:

- flagging/unflagging posts using the `f` keybinding in message-select mode
- a visual indicator of a flagged post that includes `[!]` after the message in question
- a message pop-up that displays flagged posts in a list after pressing `C-f` in the main view
- internal logic that keeps track of a currently selected post in the flagged posts overlay that is currently not used for display

What we're missing is:

[ ] The ability to scroll that window if the flagged posts reach off of it
[ ] The ability to select individual flagged posts for scrolling or other operations
[ ] The ability to unflag a post from the 'flagged posts' view
[ ] The ability to jump to a flagged post in the history of a channel
[ ] Updating the flagged post view as new posts are flagged and unflagged by other clients

I think that's where we're sitting as of right now.

**Comment by UlfS** [comment]
Awesome!
One suggestion I have is to reverse the sort order in the 'flagged posts' view, i.e. latest posts first, which is also the official mattermost client's behavior.
This would make this already quite useful in this state even without support for scrolling.

**Comment by aisamanra** [comment]
Flagged post support is now merged to master as of 3f9c9ca. Post display order matches the web interface—newest posts at top—and you can scroll through them and unflag them from the overlay. The keybinding to open them up has changed, because `C-f` was overlapping with a text editing keybinding; right now, it's `M-8` (the mnemonic being that Mattermost's "flagging" is effectively the same feature as Slack's "starred" messages, and `S-8` is how you type a star.) There's also a `/flags` command that will bring up the flagged messages view.

We don't currently support "jump to this post in the scrollback", but that's going to be enough of an internal change (in terms of how we store messages) that I've made a new issue for it at #272. The rest of this feature can be used on master, and we'll have it in the next release soon!

**Comment by UlfS** [comment]
Awesome! Thank you! :)
